tests/Agency/StoreTestAPI.cpp

```diff
@@ -0,0 +1,977 @@
+  assertEqual(read(R"([["/b"]])"), R"([{"b":2}])");
+  res = write(R"([[{"/b":3},{"/a/e":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  assertEqual(read(R"([["/b"]])"), R"([{"b":2}])");
+  res = write(R"([[{"/b":3},{"/a/e":{"in":3},"/a/b/c":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  res = write(R"([[{"/b":3},{"/a/e":{"in":3},"/a/b/c":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  res = write(R"([[{"/b":3},{"/a/b/c":{"in":3},"/a/e":{"in":3}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+  res = write(R"([[{"/b":3},{"/a/b/c":{"in":3},"/a/e":{"in":2}}]])");
+  ASSERT_EQ(consensus::apply_ret_t::APPLIED, res.front());
+  assertEqual(read(R"([["/b"]])"), R"([{"b":3}])");
+  
+  // Permute order of keys and objects within precondition
+  std::map<std::string,std::string> baz {
+    {"_id", "\"5a00203e4b660989b2ae5493\""},
+    {"index", "0"},
+    {"guid", "\"7a709cc2-1479-4079-a0a3-009cbe5674f4\""},
+    {"isActive", "true"},
+    {"balance", "\"$3,072.23\""},
+    {"picture", "\"http://placehold.it/32x32\""},
+    {"age", "21"},
+    {"eyeColor", "\"green\""},
+    {"name", R"({ "first": "Durham", "last": "Duke" })"},
+    {"tags", R"(["anim","et","id","do","est",1.0,-1024,1024])"}
+  };
+  std::string const baz_text = to_json_object(baz);
+  std::string const qux {R"(["3.14159265359",3.14159265359])"};
+  std::string const foo_value {"\"bar\""};
+  std::string const localObj = R"(
+      {"foo" : )"+ foo_value + R"(,
+        "baz" : )" + baz_text + R"(,
+        "qux" : )" + qux + R"(
+      })";
+
+  res = write(
+    std::string("[[") + localObj + R"(,
+      {
+        "baz":{"old": )" + baz_text + R"(},
+        "qux":)" + qux + R"(}]])");
+  ASSERT_EQ(consensus::apply_ret_t::PRECONDITION_FAILED, res.front());
+
+  writeAndCheck("[[" + localObj + "]]");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"foo":)" + foo_value + R"(,"baz":{"old":)" + baz_text + R"(},"qux":)" + qux + "}]]");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"baz":{"old":)" + baz_text + R"(},"foo":)" + foo_value + R"(,"qux":)" + qux + R"(}]])");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"baz":{"old":)" + baz_text + R"(},"qux":)" + qux + R"(,"foo":)" + foo_value + R"(}]])");
+  writeAndCheck(
+    "[[" + localObj + R"(, {"qux":)" + qux + R"(,"baz":{"old":)" + baz_text + R"(},"foo":)" + foo_value + R"(}]])");
+
+  std::vector<std::string> localKeys;
+  std::transform(baz.begin(), baz.end(), std::back_inserter(localKeys), [](auto kv){ return kv.first; });
+  auto rd {std::random_device{}()};
+  auto g {std::mt19937(rd)};
+  for (int permutation_count{}; permutation_count < 5; ++permutation_count) {
+    std::shuffle(localKeys.begin(), localKeys.end(), g);
+    std::string const permuted = to_json_object(localKeys, 
+      [](std::string const &k){ return k; },
```

Comment:
Another issue of style: For me, the type of `k` is `std::string const&` and therefore I put the `&` next to `const` and not next to `k`. Same for pointers, I would always write `int* x;` and not `int *x;`. I know, many do not do this, but here you go. And I would always refrain from doing things like:
```
  int *x, y;
```
We used to have coding guidelines in this way, but I do not know if they are still in force.

Reply:
Actually, I used to follow your convention (which is consistent with MFC, which occuppied my first decade of C++ practice). Then I noticed someone who I fail to remember who it was, around 2008, doing this ```std::string const &``` which at first seemed strange. At some point he gave an explanation that I thought was very smart, and started following this style naturally. Let me try to find who this person was, and what was the specific motiviation that made it better to write like that.